File: clients/python/nexus.py
"""
NexusDB Python Client Library

Usage:
    from nexus import NexusClient, query
    
    client = NexusClient("localhost:5433")
    
    # Using NQL (unified protocol)
    result = client.execute({
        "op": "select",
        "from": "users",
        "where": {"field": "age", "gt": 25},
        "limit": 10
    })
    
    # Or using context manager for transactions
    with client.transaction(isolation="serializable"):
        client.insert_one("users", {"name": "Alice", "age": 30})
        client.insert_one("users", {"name": "Bob", "age": 25})
"""

import json
import socket
from typing import Any, Dict, List, Optional, Union
from dataclasses import dataclass, asdict


@dataclass
class NQLError:
    """NQL error response"""
    code: str
    message: str
    context: Dict[str, Any]
    
    @classmethod
    def from_dict(cls, data: Dict) -> "NQLError":
        return cls(
            code=data.get("code", "UNKNOWN_ERROR"),
            message=data.get("message", "Unknown error"),
            context=data.get("context", {})
        )


@dataclass  
class QueryResult:
    """Query result wrapper"""
    rows: List[Dict[str, Any]]
    columns: List[str]
    row_count: int
    execution_time_ms: float
    
    @classmethod
    def from_response(cls, data: Dict) -> "QueryResult":
        return cls(
            rows=data.get("rows", []),
            columns=data.get("columns", []),
            row_count=data.get("row_count", 0),
            execution_time_ms=data.get("execution_time_ms", 0.0)
        )


class NexusClient:
    """NexusDB TCP client for NQL protocol"""
    
    def __init__(self, host: str = "localhost", port: int = 5433, 
                 timeout: int = 30):
        """
        Initialize NexusDB client
        
        Args:
            host: Server hostname
            port: TCP port (default 5433, NOT 5432)
            timeout: Connection timeout in seconds
        """
        self.host = host
        self.port = port
        self.timeout = timeout
        self._socket = None
        self._in_transaction = False
    
    def connect(self) -> None:
        """Connect to NexusDB server"""
        if self._socket:
            return
        
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._socket.settimeout(self.timeout)
        self._socket.connect((self.host, self.port))
    
    def disconnect(self) -> None:
        """Disconnect from server"""
        if self._socket:
            self._socket.close()
            self._socket = None
    
    def _send_command(self, command: Dict[str, Any]) -> Dict[str, Any]:
        """Send NQL command and receive response"""
        if not self._socket:
            self.connect()
        
        # Send as JSON line
        json_str = json.dumps(command)
        self._socket.send(json_str.encode() + b"\n")
        
        # Receive response (also JSON line)
        response_bytes = b""
        while b"\n" not in response_bytes:
            chunk = self._socket.recv(4096)
            if not chunk:
                raise ConnectionError("Server closed connection")
            response_bytes += chunk
        
        response = json.loads(response_bytes.decode().strip())
        
        # Check for errors
        if not response.get("success", False):
            error_data = response.get("error", {})
            raise NQLError.from_dict(error_data)
        
        return response.get("result", {})
    
    def execute(self, command: Dict[str, Any]) -> Union[QueryResult, Any]:
        """Execute NQL command"""
        result = self._send_command(command)
        
        # Return QueryResult for select operations
        if isinstance(result, dict) and "rows" in result:
            return QueryResult.from_response(result)
        
        return result
    
    def select(self, table: str, columns: Optional[List[str]] = None,
               where: Optional[Dict] = None, limit: int = None,
               offset: int = None) -> QueryResult:
        """Execute SELECT query"""
        command = {
            "op": "select",
            "from": table
        }
        
        if columns:
            command["columns"] = columns
        if where:
            command["where"] = where
        if limit:
            command["limit"] = limit
        if offset:
            command["offset"] = offset
        
        return self.execute(command)
    
    def insert(self, table: str, values: Union[Dict, List[Dict]]) -> Dict:
        """Insert one or multiple rows"""
        if isinstance(values, dict):
            values = [values]
        
        command = {
            "op": "insert",
            "into": table,
            "values": values
        }
        
        return self.execute(command)
    
    def insert_one(self, table: str, value: Dict) -> Dict:
        """Insert single row"""
        return self.insert(table, value)
    
    def update(self, table: str, updates: Dict, where: Dict) -> Dict:
        """Update rows"""
        command = {
            "op": "update",
            "table": table,
            "set": updates,
            "where": where
        }
        
        return self.execute(command)
    
    def delete(self, table: str, where: Dict) -> Dict:
        """Delete rows"""
        command = {
            "op": "delete",
            "from": table,
            "where": where
        }
        
        return self.execute(command)
    
    def create_table(self, name: str, columns: List[Dict]) -> Dict:
        """Create new table"""
        command = {
            "op": "create_table",
            "name": name,
            "columns": columns
        }
        
        return self.execute(command)
    
    def begin_transaction(self, isolation: str = "read_committed") -> None:
        """Begin transaction"""
        self._in_transaction = True
        command = {
            "op": "begin_transaction",
            "isolation_level": isolation
        }
        self.execute(command)
    
    def commit(self) -> None:
        """Commit transaction"""
        if not self._in_transaction:
            raise RuntimeError("No active transaction")
        
        self.execute({"op": "commit"})
        self._in_transaction = False
    
    def rollback(self) -> None:
        """Rollback transaction"""
        if not self._in_transaction:
            raise RuntimeError("No active transaction")
        
        self.execute({"op": "rollback"})
        self._in_transaction = False
    
    def transaction(self, isolation: str = "serializable"):
        """Context manager for transactions"""
        return TransactionContext(self, isolation)
    
    def __enter__(self):
        self.connect()
        return self
    
    def __exit__(self, *args):
        self.disconnect()


class TransactionContext:
    """Context manager for database transactions"""
    
    def __init__(self, client: NexusClient, isolation: str):
        self.client = client
        self.isolation = isolation
    
    def __enter__(self):
        self.client.begin_transaction(self.isolation)
        return self.client
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_type:
            self.client.rollback()
        else:
            self.client.commit()


# ========== Query Builder ==========

class SelectBuilder:
    """Fluent query builder for SELECT"""
    
    def __init__(self, table: str):
        self.query = {
            "op": "select",
            "from": table
        }
    
    def columns(self, *cols: str) -> "SelectBuilder":
        self.query["columns"] = list(cols)
        return self
    
    def where(self, field: str, op: str, value: Any) -> "SelectBuilder":
        self.query["where"] = {"field": field, op: value}
        return self
    
    def and_(self, field: str, op: str, value: Any) -> "SelectBuilder":
        if "where" not in self.query:
            raise ValueError("Must call where() first")
        
        self.query["where"] = {
            "and": [
                self.query["where"],
                {"field": field, op: value}
            ]
        }
        return self
    
    def or_(self, field: str, op: str, value: Any) -> "SelectBuilder":
        if "where" not in self.query:
            raise ValueError("Must call where() first")
        
        self.query["where"] = {
            "or": [
                self.query["where"],
                {"field": field, op: value}
            ]
        }
        return self
    
    def order_by(self, field: str, desc: bool = False) -> "SelectBuilder":
        self.query["order_by"] = [{"field": field, "desc": desc}]
        return self
    
    def limit(self, n: int) -> "SelectBuilder":
        self.query["limit"] = n
        return self
    
    def offset(self, n: int) -> "SelectBuilder":
        self.query["offset"] = n
        return self
    
    def build(self) -> Dict:
        return self.query


# Example usage
if __name__ == "__main__":
    # Connect to NexusDB
    with NexusClient("localhost", 5433) as client:
        # Create table
        client.create_table("users", [
            {"name": "id", "type": "integer", "primary_key": True},
            {"name": "name", "type": "string"},
            {"name": "age", "type": "integer"}
        ])
        
        # Insert data
        client.insert("users", {
            "id": 1,
            "name": "Alice",
            "age": 30
        })
        
        # Query with transaction
        with client.transaction("serializable"):
            result = client.select("users", where={"field": "age", "gt": 25})
            print(f"Found {result.row_count} users")
            for row in result.rows:
                print(row)
            
            client.insert("users", {"id": 2, "name": "Bob", "age": 28})

File: clients/php/nexus.php
<?php

/**
 * NexusDB PHP Client Library
 * 
 * Usage:
 *   $client = new NexusClient('localhost', 5433);
 *   $result = $client->select('users', where: ['field' => 'age', 'gt' => 25]);
 *   foreach ($result->rows as $row) {
 *       echo $row['name'];
 *   }
 */

namespace NexusDB;

class NexusException extends \Exception {}

class NQLError {
    public string $code;
    public string $message;
    public array $context;
    
    public function __construct(array $data) {
        $this->code = $data['code'] ?? 'UNKNOWN_ERROR';
        $this->message = $data['message'] ?? 'Unknown error';
        $this->context = $data['context'] ?? [];
    }
}

class QueryResult {
    public array $rows;
    public array $columns;
    public int $row_count;
    public float $execution_time_ms;
    
    public function __construct(array $data) {
        $this->rows = $data['rows'] ?? [];
        $this->columns = $data['columns'] ?? [];
        $this->row_count = $data['row_count'] ?? 0;
        $this->execution_time_ms = $data['execution_time_ms'] ?? 0.0;
    }
    
    public function first(): ?array {
        return $this->rows[0] ?? null;
    }
    
    public function all(): array {
        return $this->rows;
    }
}

class NexusClient {
    private string $host;
    private int $port;
    private int $timeout;
    private $socket = null;
    private bool $inTransaction = false;
    
    public function __construct(
        string $host = 'localhost',
        int $port = 5433,
        int $timeout = 30
    ) {
        $this->host = $host;
        $this->port = $port;
        $this->timeout = $timeout;
    }
    
    /**
     * Connect to NexusDB
     */
    public function connect(): void {
        if ($this->socket !== null) {
            return;
        }
        
        $this->socket = @fsockopen($this->host, $this->port, $errno, $errstr, $this->timeout);
        
        if (!$this->socket) {
            throw new NexusException("Failed to connect: $errstr ($errno)");
        }
        
        stream_set_timeout($this->socket, $this->timeout);
    }
    
    /**
     * Disconnect from server
     */
    public function disconnect(): void {
        if ($this->socket) {
            fclose($this->socket);
            $this->socket = null;
        }
    }
    
    /**
     * Send NQL command
     */
    private function sendCommand(array $command): array {
        if ($this->socket === null) {
            $this->connect();
        }
        
        $json = json_encode($command) . "\n";
        
        if (fwrite($this->socket, $json) === false) {
            throw new NexusException("Failed to send command");
        }
        
        $response = trim(fgets($this->socket, 4096));
        
        if (empty($response)) {
            throw new NexusException("No response from server");
        }
        
        $data = json_decode($response, true);
        
        if ($data === null) {
            throw new NexusException("Invalid JSON response: $response");
        }
        
        if (!$data['success'] ?? false) {
            $error = $data['error'] ?? [];
            throw new NexusException($error['message'] ?? 'Unknown error');
        }
        
        return $data['result'] ?? [];
    }
    
    /**
     * Execute NQL command
     */
    public function execute(array $command): mixed {
        $result = $this->sendCommand($command);
        
        if (isset($result['rows'])) {
            return new QueryResult($result);
        }
        
        return $result;
    }
    
    /**
     * SELECT query
     */
    public function select(
        string $table,
        ?array $columns = null,
        ?array $where = null,
        ?int $limit = null,
        ?int $offset = null
    ): QueryResult {
        $command = [
            'op' => 'select',
            'from' => $table,
        ];
        
        if ($columns !== null) {
            $command['columns'] = $columns;
        }
        if ($where !== null) {
            $command['where'] = $where;
        }
        if ($limit !== null) {
            $command['limit'] = $limit;
        }
        if ($offset !== null) {
            $command['offset'] = $offset;
        }
        
        return $this->execute($command);
    }
    
    /**
     * INSERT rows
     */
    public function insert(string $table, array $values): array {
        if (!is_array($values[0] ?? null)) {
            $values = [$values];
        }
        
        $command = [
            'op' => 'insert',
            'into' => $table,
            'values' => $values,
        ];
        
        return $this->execute($command);
    }
    
    /**
     * INSERT single row
     */
    public function insertOne(string $table, array $value): array {
        return $this->insert($table, [$value]);
    }
    
    /**
     * UPDATE rows
     */
    public function update(string $table, array $updates, array $where): array {
        $command = [
            'op' => 'update',
            'table' => $table,
            'set' => $updates,
            'where' => $where,
        ];
        
        return $this->execute($command);
    }
    
    /**
     * DELETE rows
     */
    public function delete(string $table, array $where): array {
        $command = [
            'op' => 'delete',
            'from' => $table,
            'where' => $where,
        ];
        
        return $this->execute($command);
    }
    
    /**
     * CREATE TABLE
     */
    public function createTable(string $name, array $columns): array {
        $command = [
            'op' => 'create_table',
            'name' => $name,
            'columns' => $columns,
        ];
        
        return $this->execute($command);
    }
    
    /**
     * BEGIN TRANSACTION
     */
    public function beginTransaction(string $isolation = 'read_committed'): void {
        $this->inTransaction = true;
        $this->execute([
            'op' => 'begin_transaction',
            'isolation_level' => $isolation,
        ]);
    }
    
    /**
     * COMMIT
     */
    public function commit(): void {
        if (!$this->inTransaction) {
            throw new NexusException('No active transaction');
        }
        
        $this->execute(['op' => 'commit']);
        $this->inTransaction = false;
    }
    
    /**
     * ROLLBACK
     */
    public function rollback(): void {
        if (!$this->inTransaction) {
            throw new NexusException('No active transaction');
        }
        
        $this->execute(['op' => 'rollback']);
        $this->inTransaction = false;
    }
    
    /**
     * Execute callback within transaction
     */
    public function transaction(callable $callback, string $isolation = 'serializable'): mixed {
        try {
            $this->beginTransaction($isolation);
            $result = $callback($this);
            $this->commit();
            return $result;
        } catch (\Exception $e) {
            if ($this->inTransaction) {
                $this->rollback();
            }
            throw $e;
        }
    }
    
    /**
     * Close connection on destruct
     */
    public function __destruct() {
        $this->disconnect();
    }
}

/**
 * Query builder
 */
class SelectBuilder {
    private array $query;
    
    public function __construct(string $table) {
        $this->query = [
            'op' => 'select',
            'from' => $table,
        ];
    }
    
    public function columns(string ...$cols): self {
        $this->query['columns'] = $cols;
        return $this;
    }
    
    public function where(string $field, string $operator, mixed $value): self {
        $this->query['where'] = ['field' => $field, $operator => $value];
        return $this;
    }
    
    public function and(string $field, string $operator, mixed $value): self {
        if (!isset($this->query['where'])) {
            throw new NexusException('Must call where() first');
        }
        
        $this->query['where'] = [
            'and' => [
                $this->query['where'],
                ['field' => $field, $operator => $value],
            ],
        ];
        return $this;
    }
    
    public function or(string $field, string $operator, mixed $value): self {
        if (!isset($this->query['where'])) {
            throw new NexusException('Must call where() first');
        }
        
        $this->query['where'] = [
            'or' => [
                $this->query['where'],
                ['field' => $field, $operator => $value],
            ],
        ];
        return $this;
    }
    
    public function orderBy(string $field, bool $desc = false): self {
        $this->query['order_by'] = [['field' => $field, 'desc' => $desc]];
        return $this;
    }
    
    public function limit(int $n): self {
        $this->query['limit'] = $n;
        return $this;
    }
    
    public function offset(int $n): self {
        $this->query['offset'] = $n;
        return $this;
    }
    
    public function build(): array {
        return $this->query;
    }
}

// Example usage
function example() {
    $client = new NexusClient('localhost', 5433);
    
    try {
        // Create table
        $client->createTable('users', [
            ['name' => 'id', 'type' => 'integer', 'primary_key' => true],
            ['name' => 'name', 'type' => 'string'],
            ['name' => 'age', 'type' => 'integer'],
        ]);
        
        // Insert with transaction
        $client->transaction(function($cli) {
            $cli->insert('users', [
                'id' => 1,
                'name' => 'Alice',
                'age' => 30,
            ]);
            
            $result = $cli->select('users', where: ['field' => 'age', 'gt' => 25]);
            echo "Found {$result->row_count} users\n";
            foreach ($result->rows as $row) {
                echo "{$row['name']}: {$row['age']} years\n";
            }
        });
    } finally {
        $client->disconnect();
    }
}
